httpcore/_backends/base.py

```diff
@@ -16,6 +20,17 @@ def read(self, max_bytes: int, timeout: typing.Optional[float] = None) -> bytes:
     def write(self, buffer: bytes, timeout: typing.Optional[float] = None) -> None:
         raise NotImplementedError()  # pragma: nocover
 
+    def __enter__(self) -> "Self":
```

Comment:
I don't think we need the `typing_extensions.Self` here. (?)
Can we just have this return `NetworkStream`.

The override point is `close()`, not the `__enter__`/`__exit__` which will stay the same even for subclasses.

Reply:
If we can inherit, I believe we should always use Self to avoid strange type issues, such as when the instance of SlowNetworkStream is NetworkStream.

Reply:
Example:

```
import typing
from httpcore._backends.base import NetworkStream
from time import sleep

class SlowNetworkStream(NetworkStream):
    
    def read(self, max_bytes: int, timeout: float | None = None) -> bytes:
        sleep(100)

with SlowNetworkStream() as stream:
    reveal_type(stream)
```

OUTPUT test.py:11: note: Revealed type is "httpcore._backends.base.NetworkStream"

Reply:
Ah okay right.
Could we use the `TypeVar` style, then?

Eg... in `httpx`

https://github.com/encode/httpx/blob/76c9cb65f2a159adb764c2236d139f85b46e1506/httpx/_client.py#L60
https://github.com/encode/httpx/blob/76c9cb65f2a159adb764c2236d139f85b46e1506/httpx/_client.py#L1263

Really prefer us avoiding introducing new third party packages wherever possible.

Reply:
See https://github.com/encode/httpcore/blob/e31572e0371557d6163d2b7c28676e2b1727673b/httpcore/_backends/base.py

Initially, we used TypeVar, but it was too complicated.